### ISSUE TITLE
Compilation.GetTypeByMetadataName does not account for accessibility

### ISSUE THREAD

**Issue by sharwell** [dotnet/runtime#57349]
The following code is potentially problematic:

https://github.com/dotnet/runtime/blob/c8a658ccadf808490b206d31d02c957f549d53a2/src/libraries/System.Text.Json/gen/Reflection/MetadataLoadContextInternal.cs#L26

If _any_ assembly referenced by the compilation defines a second copy of a type as `internal` (e.g. how Microsoft.CodeAnalysis defines its own copy of the [nullable attributes](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/InternalUtilities/NullableAttributes.cs)), `GetTypeByMetadataName` will return null for that type even if only one copy is accessible to the compilation. The problem occurred enough times that dotnet/roslyn-analyzers banned direct calls to `Compilation.GetTypeByMetadataName` outright and provided an alternative that [falls back to an accessibility-aware check](https://github.com/dotnet/roslyn-analyzers/blob/d0663cca11722045b9aeabb5d0d5335042084e13/src/Utilities/Compiler/WellKnownTypeProvider.cs#L81-L91).

See also https://github.com/dotnet/roslyn/issues/52399

**Comment by ericstj** [comment]
Removing @layomia assignment as he is out at the moment.  We need to make this fix for 6.0 as it would be a blocker for source generator usage.

**Comment by layomia** [comment]
Fixed for 6.0 in https://github.com/dotnet/runtime/pull/58766.

**Comment by ericstj** [comment]
@layomia @maryamariyan was this fixed in the logging generator as well?  If not, do we have an issue tracking that?

**Comment by ericstj** [comment]
Reopening because this doesn't appear to be addressed in the Logging generator.
https://github.com/dotnet/runtime/blob/73c0f8d7f6635d589470c7ff7d9680128342aac4/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs#L68-L94

**Comment by steveharter** [comment]
@sharwell this appears to only affect framework types that are redefined\cross-compiled, and not user-defined types since conflicts appear to be avoided by selecting the type in the first referenced assembly, whether the Type is `public` or not.

Since System.Text.Json resolves many internal types (all supported collection types and unsupported types) it makes sense to do that just in case. However, for other generators that use `GetTypeByMetadataName()` they just have a limited number of hard-coded cases that are unlikely to conflict. Is it still important to change these to use `GetBestTypeByMetadataName()`? Here's a list of remaining:

```
Microsoft.Extensions.Logging.Abstractions:
INamedTypeSymbol loggerMessageAttribute = _compilation.GetTypeByMetadataName(LoggerMessageAttribute);	C:\git\runtime1\src\libraries\Microsoft.Extensions.Logging.Abstractions\gen\LoggerMessageGenerator.Parser.cs	68	72
INamedTypeSymbol loggerSymbol = _compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.ILogger");	C:\git\runtime1\src\libraries\Microsoft.Extensions.Logging.Abstractions\gen\LoggerMessageGenerator.Parser.cs	75	62
INamedTypeSymbol logLevelSymbol = _compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.LogLevel");	C:\git\runtime1\src\libraries\Microsoft.Extensions.Logging.Abstractions\gen\LoggerMessageGenerator.Parser.cs	82	64
INamedTypeSymbol exceptionSymbol = _compilation.GetTypeByMetadataName("System.Exception");	C:\git\runtime1\src\libraries\Microsoft.Extensions.Logging.Abstractions\gen\LoggerMessageGenerator.Parser.cs	89	65

System.Diagnostics.Tracing:
INamedTypeSymbol? autogenerateAttribute = _compilation.GetTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAutoGenerateAttribute");	C:\git\runtime1\src\libraries\System.Private.CoreLib\generators\EventSourceGenerator.Parser.cs	34	72
INamedTypeSymbol? eventSourceAttribute = _compilation.GetTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAttribute");	C:\git\runtime1\src\libraries\System.Private.CoreLib\generators\EventSourceGenerator.Parser.cs	41	71
```

**Comment by steveharter** [comment]
> Since System.Text.Json resolves many internal types (all supported collection types and unsupported types) it makes sense to do that just in case.

Actually the collection and unsupported types are not affected by usage of `GetTypeByMetadataName` since they are not defined in STJ.

**Comment by sharwell** [comment]
I would typically use the alternative method everywhere. I can't think of a downside of this, and it's the approach used by dotnet/roslyn-analyzers.

**Comment by ericstj** [comment]
> unlikely to conflict

I bet @jaredpar can think of cases to counter this.  Wouldn't be surprised if they are in our own codebases :)

**Comment by jaredpar** [comment]
The way I usually frame this is:

> When a customer comes to you and says "I can't use your source generator because a library I reference re-defined some types" will you be happy with the response of "Sorry, you just can't use our source generator".

That is essentially the situation you can get into here. 

In some ways it's not completely unreasonable. For instance do we want to bend over backwards to support libraries that redefine `System.Exception`? I can see a stance of "no" being completely reasonable there. But it's all about your tolerance level here. 

I just want to make sure that the consequences and impact of the decisions is understood :smile:

**Comment by danmoseley** [comment]
@mkArtakMSFT are there issues in ASP.NET's usage? https://grep.app/search?current=2&q=GetTypeByMetadataName&filter[repo][0]=dotnet/aspnetcore

**Comment by danmoseley** [comment]
ping @mkArtakMSFT 😄

**Comment by ericstj** [comment]
I believe this can be closed now since #59774 was merged.   @danmoseley @mkArtakMSFT I made a new issue in ASP.NET to evaluate.
